### PR TITLE
Make type checking compatible with webpack 5

### PR DESCRIPTION
**Pull request by timneutkens**
Ensures our typescript setup does not fail from missing types / conflicts between installed tapable, webpack and definitelytyped

### CODE REVIEW

File: packages/next/server/on-demand-entry-handler.ts [hunk 1]
```diff
@@ -296,6 +296,7 @@ class Invalidator {
     // Work around a bug in webpack, calling `invalidate` on Watching.js
     // doesn't trigger the invalid call used to keep track of the `.done` hook on multiCompiler
     for (const compiler of this.multiCompiler.compilers) {
+      // @ts-ignore TODO: Check if this is still needed with webpack 5
```
Comment (sokra) [inline packages/next/server/on-demand-entry-handler.ts:299]: https://github.com/webpack/webpack/pull/11177

**Review by Timer — DISMISSED** [review]

File: packages/react-refresh-utils/ReactRefreshWebpackPlugin.ts [hunk 2]
```diff
@@ -11,8 +11,7 @@ import {
 
 // Shared between webpack 4 and 5:
 function injectRefreshFunctions(compilation: Compilation.Compilation) {
-  const hookVars: typeof compilation['mainTemplate']['hooks']['requireExtensions'] = (compilation
-    .mainTemplate.hooks as any).localVars
+  const hookVars: any = (compilation.mainTemplate.hooks as any).localVars
 
   hookVars.tap('ReactFreshWebpackPlugin', (source) =>
```
Comment (Timer) [inline packages/react-refresh-utils/ReactRefreshWebpackPlugin.ts:16]:
```suggestion
  hookVars.tap('ReactFreshWebpackPlugin', (source: string) =>
```

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| buildDuration | 13.8s | 14.5s | ⚠️ +731ms |
| nodeModulesSize | 66.5 MB | 66.5 MB | ⚠️ +51 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.321 | 2.383 | ⚠️ +0.06 |
| / avg req/sec | 1077.19 | 1049.18 | ⚠️ -28.01 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.482 | 1.517 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 1686.82 | 1647.76 | ⚠️ -39.06 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.72 kB | 6.72 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..cead.js gzip | 9.77 kB | 9.77 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.4 kB | 56.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.8 kB | 5.8 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 6.17 kB | 6.17 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.9 kB | 51.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| _buildManifest.js gzip | 267 B | 267 B | ✓ |
| _buildManife..dule.js gzip | 273 B | 273 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| index.html gzip | 954 B | 954 B | ✓ |
| link.html gzip | 962 B | 962 B | ✓ |
| withRouter.html gzip | 948 B | 948 B | ✓ |
| Overall change | 2.86 kB | 2.86 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| buildDuration | 15.1s | 15.4s | ⚠️ +263ms |
| nodeModulesSize | 66.5 MB | 66.5 MB | ⚠️ +51 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.72 kB | 6.72 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..cead.js gzip | 9.77 kB | 9.77 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.4 kB | 56.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.8 kB | 5.8 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 6.17 kB | 6.17 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.9 kB | 51.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| _buildManifest.js gzip | 267 B | 267 B | ✓ |
| _buildManife..dule.js gzip | 273 B | 273 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| _error.js | 882 kB | 882 kB | ✓ |
| 404.html | 4.17 kB | 4.17 kB | ✓ |
| hooks.html | 3.79 kB | 3.79 kB | ✓ |
| index.js | 882 kB | 882 kB | ✓ |
| link.js | 924 kB | 924 kB | ✓ |
| routerDirect.js | 918 kB | 918 kB | ✓ |
| withRouter.js | 918 kB | 918 kB | ✓ |
| Overall change | 4.53 MB | 4.53 MB | ✓ |

</details>

</details>
Commit: 4b30173670625a3c886eca85cf10b6fea68c3609

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| buildDuration | 13.7s | 13.4s | -298ms |
| nodeModulesSize | 66.5 MB | 66.5 MB | ⚠️ +51 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.289 | 2.34 | ⚠️ +0.05 |
| / avg req/sec | 1092.27 | 1068.16 | ⚠️ -24.11 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.557 | 1.585 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 1605.46 | 1577.41 | ⚠️ -28.05 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.72 kB | 6.72 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..cead.js gzip | 9.77 kB | 9.77 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.4 kB | 56.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.8 kB | 5.8 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 6.17 kB | 6.17 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.9 kB | 51.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| _buildManifest.js gzip | 267 B | 267 B | ✓ |
| _buildManife..dule.js gzip | 273 B | 273 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| index.html gzip | 954 B | 954 B | ✓ |
| link.html gzip | 962 B | 962 B | ✓ |
| withRouter.html gzip | 948 B | 948 B | ✓ |
| Overall change | 2.86 kB | 2.86 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| buildDuration | 14.2s | 14.2s | ⚠️ +14ms |
| nodeModulesSize | 66.5 MB | 66.5 MB | ⚠️ +51 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.72 kB | 6.72 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..cead.js gzip | 9.77 kB | 9.77 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.4 kB | 56.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.8 kB | 5.8 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 6.17 kB | 6.17 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.9 kB | 51.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| _buildManifest.js gzip | 267 B | 267 B | ✓ |
| _buildManife..dule.js gzip | 273 B | 273 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js webpack5-tests | Change |
| - | - | - | - |
| _error.js | 882 kB | 882 kB | ✓ |
| 404.html | 4.17 kB | 4.17 kB | ✓ |
| hooks.html | 3.79 kB | 3.79 kB | ✓ |
| index.js | 882 kB | 882 kB | ✓ |
| link.js | 924 kB | 924 kB | ✓ |
| routerDirect.js | 918 kB | 918 kB | ✓ |
| withRouter.js | 918 kB | 918 kB | ✓ |
| Overall change | 4.53 MB | 4.53 MB | ✓ |

</details>

</details>
Commit: 09271941008900e23ff77125b91b744fd0697c1a

**Pull request merged**